### ISSUE TITLE
Validation loss get worse in lmptb_10k

### ISSUE THREAD

**Issue by ricky820422** [tensorflow/tensor2tensor#108]
I'm training on a single nvidia Tesla K40m with the following parameters: 
PROBLEM=lmptb_10k
MODEL=attention_lm_moe
HPARAMS=attention_lm_moe_small

As the training progressed, loss on the training set dropped to around 0.3. However, validation loss kept getting worse from 6.93 to 9.74.

`INFO:tensorflow:Validation (step 2000): loss = 6.93099, metrics-lmptb_10k/accuracy_top5 = 0.382969, global_step = 1967, metrics/neg_log_perplexity = -6.91322, metrics-lmptb_10k/accuracy_per_sequence = 0.0, metrics/accuracy = 0.215393, metrics-lmptb_10k/accuracy = 0.215393, metrics/accuracy_per_sequence = 0.0, metrics/accuracy_top5 = 0.382969, metrics-lmptb_10k/neg_log_perplexity = -6.91322`
`INFO:tensorflow:Validation (step 4000): loss = 8.7972, metrics-lmptb_10k/accuracy_top5 = 0.362099, global_step = 3904, metrics/neg_log_perplexity = -8.78079, metrics-lmptb_10k/accuracy_per_sequence = 0.0, metrics/accuracy = 0.214791, metrics-lmptb_10k/accuracy = 0.214791, metrics/accuracy_per_sequence = 0.0, metrics/accuracy_top5 = 0.362099, metrics-lmptb_10k/neg_log_perplexity = -8.78079`
`INFO:tensorflow:Validation (step 6000): loss = 9.38008, metrics-lmptb_10k/accuracy_top5 = 0.35822, global_step = 5838, metrics/neg_log_perplexity = -9.35931, metrics-lmptb_10k/accuracy_per_sequence = 0.0, metrics/accuracy = 0.214106, metrics-lmptb_10k/accuracy = 0.214106, metrics/accuracy_per_sequence = 0.0, metrics/accuracy_top5 = 0.35822, metrics-lmptb_10k/neg_log_perplexity = -9.35931`
`INFO:tensorflow:Validation (step 8000): loss = 9.74558, metrics-lmptb_10k/accuracy_top5 = 0.354445, global_step = 7766, metrics/neg_log_perplexity = -9.72227, metrics-lmptb_10k/accuracy_per_sequence = 0.0, metrics/accuracy = 0.214293, metrics-lmptb_10k/accuracy = 0.214293, metrics/accuracy_per_sequence = 0.0, metrics/accuracy_top5 = 0.354445, metrics-lmptb_10k/neg_log_perplexity = -9.72227`

Any suggestions to make it work? Do I need to tune the parameters by myself?

Also, I always got “out of range” warning while performing evaluation as follows:

`INFO:tensorflow:Starting evaluation at 2017-07-06-07:32:07`
`2017-07-06 07:32:09.066315: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K40m, pci bus id: 0000:84:00.0)`
`INFO:tensorflow:Restoring parameters from /notebooks/share/output/t2t_train/lmptb_10k/attention_lm_moe-attention_lm_moe_small/model.ckpt-7766`
`INFO:tensorflow:Evaluation [1/10]`
`INFO:tensorflow:Evaluation [2/10]`
`INFO:tensorflow:Evaluation [3/10]`
`INFO:tensorflow:Evaluation [4/10]`
`INFO:tensorflow:Evaluation [5/10]`
`INFO:tensorflow:Evaluation [6/10]`
`INFO:tensorflow:Evaluation [7/10]`
`2017-07-06 07:32:18.685524: W tensorflow/core/framework/op_kernel.cc:1158] Out of range: FIFOQueue '_64_input_queues/problem_0/batch_examples/bucket_by_sequence_length/bucket/top_queue' is closed and has insufficient elements (requested 1, current size 0)`
         `[[Node: input_queues/problem_0/batch_examples/bucket_by_sequence_length/bucket/dequeue_top = QueueDequeueV2[component_types=[DT_INT32, DT_INT32, DT_INT32, DT_INT32], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/cpu:0"](input_queues/problem_0/batch_examples/bucket_by_sequence_length/bucket/top_queue)]]`
`INFO:tensorflow:Finished evaluation at 2017-07-06-07:32:19`

Is there something wrong with the evaluation process?
Thanks a lot!

**Comment by lukaszkaiser** [comment]
This is no surprise to me: you're running a moe=Mixture-of-Experts model (mixtures of experts are models with large number of parameters to fit large datasets) on a very small dataset, the PTB. The eval failure is probably due to the fact that the eval set is too small and you're running out of it. I'd suggest you run the non-MoE model, and maybe in a smaller setting. The MoE model was tuned for the 1B words benchmark, for PTB you need to tune the parameters yourself.

**Comment by ricky820422** [comment]
@lukaszkaiser 
Thank you very much for your detailed explanation.
I have tried the non-MoE model, like the following parameters:
PROBLEM=lmptb_10k
MODEL=attention_lm
HPARAMS=attention_lm_base

Still got similar results, so I tried to turn down the hidden_size and num_hidden_layers, also increased the attention_dropout and residual_dropout, but still no satisfactory results with log_perplexity around 5.00.

What's the best result you've got on PTB dataset? any other parameters I need to tune? or any further advice for my situation?

Thanks a lot.

**Comment by lukaszkaiser** [comment]
You definitely need to lower `filter_size` and `hidden_size` and set all dropouts higher. Also, as far as I can tell, log-ppx of `4.4` is quite good and `4.1` would be state-of-the-art, so `5.0` isn't that bad for a model that's totally untuned for the task. Note that we never ran our models on PTB as it's a very small data-set and it was only added recently in a github PR. I'm just giving it a shot with these params:

--problems=lmptb_10k   --model=attention_lm   --hparams_set=attention_lm_base  --train_steps=20000   --eval_steps=7 --hparams='hidden_size=512,filter_size=1024,dropout=0.5,attention_dropout=0.5,label_smoothing=0.1,relu_dropout=0.5,residual_dropout=0.5,learning_rate_warmup_steps=2000,learning_rate_decay_scheme=exp10k'

I'm getting `4.8` after 8k steps, so not too horrible, but still way to go!

If you get a better hparams set for ptb, please make a PR adding it to `attention_lm` as `attention_lm_ptb` :). It's also sth to collaborate on with others!

**Comment by ricky820422** [comment]
@lukaszkaiser Thanks! Working on that!
